File: samples_python/ecs_sample.py
import json
import time
from samples_python.task_status import TaskStatus
from alibabacloud_cloudcontrol20220606.client import Client as CloudControlClient
from alibabacloud_tea_openapi import models as open_api_models
from alibabacloud_cloudcontrol20220606 import models as cloud_control_models
from alibabacloud_tea_util.client import Client as UtilClient


class EcsSample:

    def __init__(self):
        pass

    @staticmethod
    def create_client(
    ) -> CloudControlClient:
        """
        使用AK&SK初始化账号Client
        @return: Client
        @throws Exception
        """
        config = open_api_models.Config(
            # 您的AccessKey ID(改为自己的)
            access_key_id="your_ak",
            # 您的AccessKey Secret(改为自己的)
            access_key_secret="your_sk",
            # 公司UserAgent信息(改为自己的)
            user_agent="vendor/AnyCloudCompany-CloudManager@2.1.0",
            # 设置sdk读取超时时间
            read_timeout=20000
        )
        # 访问的域名
        config.endpoint = 'cloudcontrol.aliyuncs.com'
        return CloudControlClient(config)

    @staticmethod
    def create_instance(
    ) -> str:
        """
        创建弹性计算实例
        @return: resourceId
        @throws Exception
        """
        # 资源属性及其对应值（json字符串,部分属性需改为自己的值）
        resource_attribute_map = {"InstanceName": "cc-test", "SystemDisk": {"Category": "cloud_efficiency"},
                                  "ZoneId": "cn-zhangjiakou-a", "SecurityGroupId": "sg-xxxxx",
                                  "ImageId": "centos_7_06_64_20G_alibase_20190711.vhd",
                                  "InstanceType": "ecs.s6-c1m1.small", "VpcAttributes": {"VSwitchId": "vsw-xxxxx"},
                                  "Tags": [{"TagKey": "cckey", "TagValue": "ccvalue"}]}
        resource_attributes = json.dumps(resource_attribute_map)

        create_resource_request = cloud_control_models.CreateResourceRequest()
        create_resource_request.region_id = regionId
        create_resource_request.body = resource_attributes
        create_resource_response = cloud_control_client.create_resource(provider, productCode, resourceTypeCode,
                                                                        create_resource_request)
        if create_resource_response.status_code == 201:
            return create_resource_response.body.resource_id
        elif create_resource_response.status_code == 202:
            start_time = time.time()
            get_task_response = None
            task_status = ''
            # 轮询异步创建任务
            while time.time() - start_time < create_resource_response.body.timeout:
                get_task_response = cloud_control_client.get_task(create_resource_response.body.task_id)
                task_status = get_task_response.body.task.status
                if task_status != TaskStatus.RUNNING.value:
                    break
                time.sleep(period)

            # 异步任务状态判断
            if task_status == TaskStatus.Succeeded.value:
                return get_task_response.body.task.resource_id
            elif task_status == TaskStatus.FAILED.value:
                error = get_task_response.body.task.error
                print("errorCode is %s, errorMessage is %s" % (error.code, error.message))
                raise Exception("Task is failed!")
            elif task_status == TaskStatus.RUNNING.value:
                cloud_control_client.cancel_task(create_resource_response.body.task_id)
                raise Exception("Task is timeout!")
            else:
                raise Exception("Invalid Task status")
        else:
            raise Exception("Invalid status_code!")

    @staticmethod
    def update_instance(
            resource_id: str
    ) -> bool:
        """
        更新弹性计算实例
        @return: bool
        @throws Exception
        """
        # 资源属性及其对应值（json字符串）
        resource_attribute_map = {'InstanceName': "cc-test2", 'InstanceType': "ecs.s6-c1m2.small",
                                  "Tags": [{"TagKey": "cckey2", "TagValue": "ccvalue2"}]}
        resource_attributes = json.dumps(resource_attribute_map)

        update_resource_request = cloud_control_models.UpdateResourceRequest()
        update_resource_request.region_id = regionId
        update_resource_request.body = resource_attributes
        update_resource_response = cloud_control_client.update_resource(provider, productCode, resourceTypeCode,
                                                                        resource_id, update_resource_request)
        if update_resource_response.status_code == 200:
            return True
        elif update_resource_response.status_code == 202:
            start_time = time.time()
            get_task_response = None
            task_status = ''
            # 轮询异步创建任务
            while time.time() - start_time < update_resource_response.body.timeout:
                get_task_response = cloud_control_client.get_task(update_resource_response.body.task_id)
                task_status = get_task_response.body.task.status
                if task_status != TaskStatus.RUNNING.value:
                    break
                time.sleep(period)

            # 异步任务状态判断
            if task_status == TaskStatus.Succeeded.value:
                return True
            elif task_status == TaskStatus.FAILED.value:
                error = get_task_response.body.task.error
                print("errorCode is %s, errorMessage is %s" % (error.code, error.message))
                raise Exception("Task is failed!")
            elif task_status == TaskStatus.RUNNING.value:
                cloud_control_client.cancel_task(update_resource_response.body.task_id)
                raise Exception("Task is timeout!")
            else:
                raise Exception("Invalid Task status")
        else:
            raise Exception("Invalid status_code!")

    @staticmethod
    def start_instance(
            resource_id: str
    ) -> bool:
        """
        启动弹性计算实例
        @return: bool
        @throws Exception
        """
        # 资源属性及其对应值（json字符串）
        resource_attribute_map = {'Status': "Running"}
        resource_attributes = json.dumps(resource_attribute_map)

        update_resource_request = cloud_control_models.UpdateResourceRequest()
        update_resource_request.region_id = regionId
        update_resource_request.body = resource_attributes
        update_resource_response = cloud_control_client.update_resource(provider, productCode, resourceTypeCode,
                                                                        resource_id, update_resource_request)
        if update_resource_response.status_code == 200:
            return True
        elif update_resource_response.status_code == 202:
            start_time = time.time()
            get_task_response = None
            task_status = ''
            # 轮询异步创建任务
            while time.time() - start_time < update_resource_response.body.timeout:
                get_task_response = cloud_control_client.get_task(update_resource_response.body.task_id)
                task_status = get_task_response.body.task.status
                if task_status != TaskStatus.RUNNING.value:
                    break
                time.sleep(period)

            # 异步任务状态判断
            if task_status == TaskStatus.Succeeded.value:
                return True
            elif task_status == TaskStatus.FAILED.value:
                error = get_task_response.body.task.error
                print("errorCode is %s, errorMessage is %s" % (error.code, error.message))
                raise Exception("Task is failed!")
            elif task_status == TaskStatus.RUNNING.value:
                cloud_control_client.cancel_task(update_resource_response.body.task_id)
                raise Exception("Task is timeout!")
            else:
                raise Exception("Invalid Task status")
        else:
            raise Exception("Invalid status_code!")

    @staticmethod
    def stop_instance(
            resource_id: str
    ) -> bool:
        """
        启动弹性计算实例
        @return: bool
        @throws Exception
        """
        # 资源属性及其对应值（json字符串）
        resource_attribute_map = {'Status': "Stopped"}
        resource_attributes = json.dumps(resource_attribute_map)

        update_resource_request = cloud_control_models.UpdateResourceRequest()
        update_resource_request.region_id = regionId
        update_resource_request.body = resource_attributes
        update_resource_response = cloud_control_client.update_resource(provider, productCode, resourceTypeCode,
                                                                        resource_id, update_resource_request)
        if update_resource_response.status_code == 200:
            return True
        elif update_resource_response.status_code == 202:
            start_time = time.time()
            get_task_response = None
            task_status = ''
            # 轮询异步创建任务
            while time.time() - start_time < update_resource_response.body.timeout:
                get_task_response = cloud_control_client.get_task(update_resource_response.body.task_id)
                task_status = get_task_response.body.task.status
                if task_status != TaskStatus.RUNNING.value:
                    break
                time.sleep(period)

            # 异步任务状态判断
            if task_status == TaskStatus.Succeeded.value:
                return True
            elif task_status == TaskStatus.FAILED.value:
                error = get_task_response.body.task.error
                print("errorCode is %s, errorMessage is %s" % (error.code, error.message))
                raise Exception("Task is failed!")
            elif task_status == TaskStatus.RUNNING.value:
                cloud_control_client.cancel_task(update_resource_response.body.task_id)
                raise Exception("Task is timeout!")
            else:
                raise Exception("Invalid Task status")
        else:
            raise Exception("Invalid status_code!")

    @staticmethod
    def get_instance(
            resource_id: str
    ) -> cloud_control_models.GetResourceResponseBodyResource:
        """
        查询弹性计算实例
        @return: GetResourceResponseBodyResource
        @throws Exception
        """
        get_resource_request = cloud_control_models.GetResourceRequest()
        get_resource_request.region_id = regionId
        get_resource_response = cloud_control_client.get_resource(provider, productCode, resourceTypeCode,
                                                                  resource_id, get_resource_request)
        if get_resource_response.status_code == 200:
            return get_resource_response.body.resource
        else:
            raise Exception("Invalid status_code!")

    @staticmethod
    def list_instances(
    ) -> list:
        """
        列举弹性计算实例
        @return: list
        @throws Exception
        """
        list_resources_request = cloud_control_models.ListResourcesRequest()
        list_resources_request.region_ids = [regionId]
        filter_map = {'InstanceName': "cc-test2"}
        list_resources_request.filter = filter_map
        list_resources_response = cloud_control_client.list_resources(provider, productCode, resourceTypeCode,
                                                                      list_resources_request)
        if list_resources_response.status_code == 200:
            return list_resources_response.body.resources
        else:
            raise Exception("Invalid status_code!")

    @staticmethod
    def delete_instance(
            resource_id: str
    ) -> bool:
        """
        删除弹性计算实例
        @return: bool
        @throws Exception
        """
        delete_resource_request = cloud_control_models.DeleteResourceRequest()
        delete_resource_request.region_id = regionId
        delete_resource_response = cloud_control_client.delete_resource(provider, productCode, resourceTypeCode,
                                                                        resource_id, delete_resource_request)
        if delete_resource_response.status_code == 200:
            return True
        elif delete_resource_response.status_code == 202:
            start_time = time.time()
            get_task_response = None
            task_status = ''
            # 轮询异步创建任务
            while time.time() - start_time < delete_resource_response.body.timeout:
                get_task_response = cloud_control_client.get_task(delete_resource_response.body.task_id)
                task_status = get_task_response.body.task.status
                if task_status != TaskStatus.RUNNING.value:
                    break
                time.sleep(period)

            # 异步任务状态判断
            if task_status == TaskStatus.Succeeded.value:
                return True
            elif task_status == TaskStatus.FAILED.value:
                error = get_task_response.body.task.error
                print("errorCode is %s, errorMessage is %s" % (error.code, error.message))
                raise Exception("Task is failed!")
            elif task_status == TaskStatus.RUNNING.value:
                cloud_control_client.cancel_task(delete_resource_response.body.task_id)
                raise Exception("Task is timeout!")
            else:
                raise Exception("Invalid Task status")
        else:
            raise Exception("Invalid status_code!")


if __name__ == '__main__':
    # 云厂商
    provider = "Aliyun"
    # 产品code
    productCode = "ECS"
    # 资源code
    resourceTypeCode = "Instance"
    # 地域
    regionId = "cn-zhangjiakou"
    # 异步操作轮询间隔(s)
    period = 1

    try:
        # 初始化sdk
        cloud_control_client = EcsSample.create_client()
        # 创建资源
        resource_id = EcsSample.create_instance()
        # 更新资源（修改名称、实例规格）
        update_success = EcsSample.update_instance(resource_id)
        # 更新资源（启动实例）
        start_success = EcsSample.start_instance(resource_id)
        # 更新资源（停止实例）
        stop_success = EcsSample.stop_instance(resource_id)
        # 查询资源
        resource = EcsSample.get_instance(resource_id)
        print(resource)
        # 列举资源
        resources = EcsSample.list_instances()
        for item in resources:
            print(item)
        # 删除资源(ecs实例创建1分钟内无法删除，故等待一段时间再删除)
        time.sleep(30)
        delete_success = EcsSample.delete_instance(resource_id)
        a=0
    except Exception as error:
        # 打印 error
        print(UtilClient.assert_as_string(error.message))

File: samples_java/src/main/java/EcsSample.java
import com.aliyun.cloudcontrol20220606.Client;
import com.aliyun.cloudcontrol20220606.models.*;
import com.aliyun.tea.TeaException;
import com.aliyun.teaopenapi.models.Config;
import com.google.gson.Gson;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author bzran
 * @description ECS实例的创建、更新、查询、列举、删除
 */
public class EcsSample {

    /**
     * 云厂商
     */
    private final static String provider = "Aliyun";

    /**
     * 云产品
     */
    private final static String productCode = "ECS";

    /**
     * 资源类型
     */
    private final static String resourceTypeCode = "Instance";

    /**
     * 地域
     */
    private final static String regionId = "cn-zhangjiakou";

    /**
     * SDK Client
     */
    private static com.aliyun.cloudcontrol20220606.Client cloudControlClient;

    /**
     * 异步操作轮询间隔(ms)
     */
    private static final int PERIOD = 1000;


    public static void main(String[] args) {

        try {
            Config config = new Config()
                    // 您的AccessKey ID
                    .setAccessKeyId("your_ak")
                    // 您的AccessKey Secret
                    .setAccessKeySecret("your_sk")
                    //设置sdk超时时间
                    .setReadTimeout(20000);
            config.endpoint = "cloudcontrol.aliyuncs.com";
            cloudControlClient = new Client(config);

            //创建ecs实例
            String instanceId = createInstance();
            //查询ecs实例
            String instance = getInstance(instanceId);
            System.out.println(instance);
            //启动ecs实例（修改运行状态属性为Running）
            boolean start = startInstance(instanceId);
            //停止ecs实例（修改运行状态属性为Stopped）
            boolean stop = stopInstance(instanceId);
            //更新ecs实例（名称、实例规格，运行状态等属性均可修改）
            boolean update = updateInstance(instanceId);
            //列举ecs实例
            String instances = listInstances();
            System.out.println(instances);
            //ecs实例创建1分钟内无法删除，故等待一段时间再删除
            Thread.sleep(30000);
            //删除ecs实例
            boolean delete = deleteInstance(instanceId);
        } catch (TeaException e) {
            // 如有需要，请打印 error
            System.out.println("Error code: " + e.getCode());
            System.out.println("Error message: " + e.getMessage());
        } catch (Exception e) {
            TeaException error = new TeaException(e.getMessage(), e);
            // 打印 error
            System.out.println(com.aliyun.teautil.Common.assertAsString(error.message));
        }
    }

    /**
     * 创建ecs实例
     */
    private static String createInstance() throws Exception {

        //资源属性及其对应值（key，value的map转为String即可,部分属性需改为自己的值）
        Map<String, Object> resourceAttributeMap = new HashMap<>();
        //自己的镜像id
        resourceAttributeMap.put("ImageId", "centos_7_06_64_20G_alibase_20190711.vhd");
        //自己的安全组id
        resourceAttributeMap.put("SecurityGroupId", "sg-xxxxx");
        //自定义可用的实例规格
        resourceAttributeMap.put("InstanceType", "ecs.s6-c1m1.small");
        //自定义的实例名称
        resourceAttributeMap.put("InstanceName", "cc-test");
        //可用区id
        resourceAttributeMap.put("ZoneId", "cn-zhangjiakou-a");
        //系统磁盘信息
        Map<String, Object> systemDisk = new HashMap<>();
        systemDisk.put("Category", "cloud_efficiency");
        resourceAttributeMap.put("SystemDisk", systemDisk);
        //VPC网络信息（自己的VSwitchId）
        Map<String, Object> VpcAttributes = new HashMap<>();
        VpcAttributes.put("VSwitchId", "vsw-xxxxx");
        resourceAttributeMap.put("VpcAttributes", VpcAttributes);
        //Tag标签
        List<Map<String, String>> tags = new ArrayList<>();
        Map<String, String> tag1 = new HashMap<>();
        tag1.put("TagKey","cckey1");
        tag1.put("TagValue","ccvalue1");
        Map<String, String> tag2 = new HashMap<>();
        tag2.put("TagKey","cckey2");
        tag2.put("TagValue","ccvalue2");
        tags.add(tag1);
        tags.add(tag2);
        resourceAttributeMap.put("Tags", tags);
        String resourceAttributes = new Gson().toJson(resourceAttributeMap);
        CreateResourceRequest createResourceRequest = new CreateResourceRequest();
        createResourceRequest.setBody(resourceAttributes);
        createResourceRequest.setRegionId(regionId);
        //返回资源的id或对应taskid
        CreateResourceResponse createResourceResponse = cloudControlClient.createResource(provider, productCode, resourceTypeCode, createResourceRequest);
        String resourceId = null;
        if (createResourceResponse.statusCode == 201) {
            resourceId = createResourceResponse.getBody().resourceId;
        } else if (createResourceResponse.statusCode == 202) {
            String taskId = createResourceResponse.getBody().getTaskId();
            long timeStart = System.currentTimeMillis();
            GetTaskResponse getTaskResponse = cloudControlClient.getTask(taskId);
            String taskStatus = getTaskResponse.getBody().getTask().status;
            // 轮询异步创建任务
            while (TaskStatusEnum.RUNNING.getValue().equals(taskStatus) && (System.currentTimeMillis() - timeStart) < createResourceResponse.getBody().getTimeout() * 1000) {
                getTaskResponse = cloudControlClient.getTask(taskId);
                taskStatus = getTaskResponse.getBody().getTask().status;
                Thread.sleep(PERIOD);
            }

            //判断异步任务结果
            if (TaskStatusEnum.Succeeded.getValue().equals(taskStatus)) {
                //成功
                resourceId = getTaskResponse.getBody().getTask().resourceId;
                System.out.println("Asyn Creating succeed.");
            } else if (TaskStatusEnum.FAILED.getValue().equals(taskStatus)) {
                //失败输出错误信息
                GetTaskResponseBody.GetTaskResponseBodyTaskError error = getTaskResponse.getBody().getTask().getError();
                System.out.println(String.format("errorCode is %s, errorMessage is %s.", error.getCode(), error.getMessage()));
                throw new Exception("Asyn Creating failed.");
            }else if (TaskStatusEnum.RUNNING.getValue().equals(taskStatus)){
                //超时则取消该任务
                cloudControlClient.cancelTask(taskId);
                throw new Exception("Task is timeout!");
            }else {
                throw new Exception("Invalid Task status!");
            }
        } else {
            throw new Exception("Invalid status code!");
        }
        return resourceId;
    }

    /**
     * 启动ecs实例（修改运行状态为Running）
     */
    private static boolean startInstance(String resourceId) throws Exception {

        //资源属性及其对应值（key，value的map转为String即可）
        Map<String, Object> resourceAttributeMap = new HashMap<>();
        //实例名称
        resourceAttributeMap.put("Status", "Running");

        String resourceAttributes = new Gson().toJson(resourceAttributeMap);
        UpdateResourceRequest updateResourceRequest = new UpdateResourceRequest();
        updateResourceRequest.setBody(resourceAttributes);
        updateResourceRequest.setRegionId(regionId);
        UpdateResourceResponse updateResourceResponse = cloudControlClient.updateResource(provider, productCode, resourceTypeCode, resourceId, updateResourceRequest);
        if (updateResourceResponse.statusCode == 200) {
            return true;
        } else if (updateResourceResponse.statusCode == 202) {
            String taskId = updateResourceResponse.getBody().getTaskId();
            long timeStart = System.currentTimeMillis();
            GetTaskResponse getTaskResponse = cloudControlClient.getTask(taskId);
            String taskStatus = getTaskResponse.getBody().getTask().status;
            // 轮询异步创建任务
            while (TaskStatusEnum.RUNNING.getValue().equals(taskStatus) && (System.currentTimeMillis() - timeStart) < updateResourceResponse.getBody().getTimeout() * 1000) {
                getTaskResponse = cloudControlClient.getTask(taskId);
                taskStatus = getTaskResponse.getBody().getTask().status;
                Thread.sleep(PERIOD);
            }

            //判断异步任务结果
            if (TaskStatusEnum.Succeeded.getValue().equals(taskStatus)) {
                //成功
                resourceId = getTaskResponse.getBody().getTask().resourceId;
                System.out.println("Asyn starting succeed.");
                return true;
            } else if (TaskStatusEnum.FAILED.getValue().equals(taskStatus)) {
                //失败输出错误信息
                GetTaskResponseBody.GetTaskResponseBodyTaskError error = getTaskResponse.getBody().getTask().getError();
                System.out.println(String.format("errorCode is %s, errorMessage is %s.", error.getCode(), error.getMessage()));
                return false;
            }else if (TaskStatusEnum.RUNNING.getValue().equals(taskStatus)){
                //超时则取消该任务
                cloudControlClient.cancelTask(taskId);
                throw new Exception("Task is timeout!");
            }else {
                throw new Exception("Invalid Task status!");
            }
        } else {
            throw new Exception("Invalid status code!");
        }
    }

    /**
     * 更新ecs实例（名称、实例规格，运行状态等属性均可修改）
     */
    private static boolean updateInstance(String resourceId) throws Exception {

        //资源属性及其对应值（key，value的map转为String即可）
        Map<String, Object> resourceAttributeMap = new HashMap<>();
        //实例名称
        resourceAttributeMap.put("InstanceName", "cc-test2");
        //实例规格
        resourceAttributeMap.put("InstanceType", "ecs.s6-c1m2.small");

        String resourceAttributes = new Gson().toJson(resourceAttributeMap);
        UpdateResourceRequest updateResourceRequest = new UpdateResourceRequest();
        updateResourceRequest.setBody(resourceAttributes);
        updateResourceRequest.setRegionId(regionId);
        UpdateResourceResponse updateResourceResponse = cloudControlClient.updateResource(provider, productCode, resourceTypeCode, resourceId, updateResourceRequest);
        if (updateResourceResponse.statusCode == 200) {
            return true;
        } else if (updateResourceResponse.statusCode == 202) {
            String taskId = updateResourceResponse.getBody().getTaskId();
            long timeStart = System.currentTimeMillis();
            GetTaskResponse getTaskResponse = cloudControlClient.getTask(taskId);
            String taskStatus = getTaskResponse.getBody().getTask().status;
            // 轮询异步创建任务
            while (TaskStatusEnum.RUNNING.getValue().equals(taskStatus) && (System.currentTimeMillis() - timeStart) < updateResourceResponse.getBody().getTimeout() * 1000) {
                getTaskResponse = cloudControlClient.getTask(taskId);
                taskStatus = getTaskResponse.getBody().getTask().status;
                Thread.sleep(PERIOD);
            }

            //判断异步任务结果
            if (TaskStatusEnum.Succeeded.getValue().equals(taskStatus)) {
                //成功
                resourceId = getTaskResponse.getBody().getTask().resourceId;
                System.out.println("Asyn updating succeed.");
                return true;
            } else if (TaskStatusEnum.FAILED.getValue().equals(taskStatus)) {
                //失败输出错误信息
                GetTaskResponseBody.GetTaskResponseBodyTaskError error = getTaskResponse.getBody().getTask().getError();
                System.out.println(String.format("errorCode is %s, errorMessage is %s.", error.getCode(), error.getMessage()));
                return false;
            }else if (TaskStatusEnum.RUNNING.getValue().equals(taskStatus)){
                //超时则取消该任务
                cloudControlClient.cancelTask(taskId);
                throw new Exception("Task is timeout!");
            }else {
                throw new Exception("Invalid Task status!");
            }
        } else {
            throw new Exception("Invalid status code!");
        }
    }

    /**
     * 停止ecs实例（修改运行状态为Stopped）
     */
    private static boolean stopInstance(String resourceId) throws Exception {

        //资源属性及其对应值（key，value的map转为String即可）
        Map<String, Object> resourceAttributeMap = new HashMap<>();
        //实例名称
        resourceAttributeMap.put("Status", "Stopped");

        String resourceAttributes = new Gson().toJson(resourceAttributeMap);
        UpdateResourceRequest updateResourceRequest = new UpdateResourceRequest();
        updateResourceRequest.setBody(resourceAttributes);
        updateResourceRequest.setRegionId(regionId);
        UpdateResourceResponse updateResourceResponse = cloudControlClient.updateResource(provider, productCode, resourceTypeCode, resourceId, updateResourceRequest);
        if (updateResourceResponse.statusCode == 200) {
            return true;
        } else if (updateResourceResponse.statusCode == 202) {
            String taskId = updateResourceResponse.getBody().getTaskId();
            long timeStart = System.currentTimeMillis();
            GetTaskResponse getTaskResponse = cloudControlClient.getTask(taskId);
            String taskStatus = getTaskResponse.getBody().getTask().status;
            // 轮询异步创建任务
            while (TaskStatusEnum.RUNNING.getValue().equals(taskStatus) && (System.currentTimeMillis() - timeStart) < updateResourceResponse.getBody().getTimeout() * 1000) {
                getTaskResponse = cloudControlClient.getTask(taskId);
                taskStatus = getTaskResponse.getBody().getTask().status;
                Thread.sleep(PERIOD);
            }

            //判断异步任务结果
            if (TaskStatusEnum.Succeeded.getValue().equals(taskStatus)) {
                //成功
                resourceId = getTaskResponse.getBody().getTask().resourceId;
                System.out.println("Asyn stopping succeed.");
                return true;
            } else if (TaskStatusEnum.FAILED.getValue().equals(taskStatus)) {
                //失败输出错误信息
                GetTaskResponseBody.GetTaskResponseBodyTaskError error = getTaskResponse.getBody().getTask().getError();
                System.out.println(String.format("errorCode is %s, errorMessage is %s.", error.getCode(), error.getMessage()));
                return false;
            }else if (TaskStatusEnum.RUNNING.getValue().equals(taskStatus)){
                //超时则取消该任务
                cloudControlClient.cancelTask(taskId);
                throw new Exception("Task is timeout!");
            }else {
                throw new Exception("Invalid Task status!");
            }
        } else {
            throw new Exception("Invalid status code!");
        }
    }

    /**
     * 查询ecs实例
     */
    private static String getInstance(String resourceId) throws Exception {

        GetResourceRequest getResourceRequest = new GetResourceRequest();
        //region化的产品必须加此参数，表示获取哪个region的资源
        getResourceRequest.setRegionId(regionId);
        //返回指定资源
        GetResourceResponse getResourceResponse = cloudControlClient.getResource(provider, productCode, resourceTypeCode, resourceId, getResourceRequest);
        return getResourceResponse.getBody().getResource().getResourceAttributes();
    }

    /**
     * 列举ecs实例
     */
    private static String listInstances() throws Exception {

        ListResourcesRequest listResourceRequest = new ListResourcesRequest();
        List<String> regionIds = new ArrayList<>();
        regionIds.add(regionId);
        listResourceRequest.setRegionIds(regionIds);
        //返回资源列表
        ListResourcesResponse listResourceResponse = cloudControlClient.listResources(provider, productCode, resourceTypeCode, listResourceRequest);
        return new Gson().toJson(listResourceResponse.getBody().getResources());
    }

    /**
     * 删除ecs实例
     */
    private static boolean deleteInstance(String resourceId) throws Exception {

        DeleteResourceRequest deleteResourceRequest = new DeleteResourceRequest();
        //region化的产品必须加此参数，表示删除哪个region的资源
        deleteResourceRequest.setRegionId(regionId);
        DeleteResourceResponse deleteResourceResponse = cloudControlClient.deleteResource(provider, productCode, resourceTypeCode, resourceId, deleteResourceRequest);
        if (deleteResourceResponse.statusCode == 200) {
            return true;
        } else if (deleteResourceResponse.statusCode == 202) {
            String taskId = deleteResourceResponse.getBody().getTaskId();
            long timeStart = System.currentTimeMillis();
            GetTaskResponse getTaskResponse = cloudControlClient.getTask(taskId);
            String taskStatus = getTaskResponse.getBody().getTask().status;
            // 轮询异步创建任务
            while (TaskStatusEnum.RUNNING.getValue().equals(taskStatus) && (System.currentTimeMillis() - timeStart) < deleteResourceResponse.getBody().getTimeout() * 1000) {
                getTaskResponse = cloudControlClient.getTask(taskId);
                taskStatus = getTaskResponse.getBody().getTask().status;
                Thread.sleep(PERIOD);
            }

            //判断异步任务结果
            if (TaskStatusEnum.Succeeded.getValue().equals(taskStatus)) {
                //成功
                resourceId = getTaskResponse.getBody().getTask().resourceId;
                System.out.println("Asyn deleting succeed.");
                return true;
            } else if (TaskStatusEnum.FAILED.getValue().equals(taskStatus)) {
                //失败输出错误信息
                GetTaskResponseBody.GetTaskResponseBodyTaskError error = getTaskResponse.getBody().getTask().getError();
                System.out.println(String.format("errorCode is %s, errorMessage is %s.", error.getCode(), error.getMessage()));
                return false;
            }else if (TaskStatusEnum.RUNNING.getValue().equals(taskStatus)){
                //超时则取消该任务
                cloudControlClient.cancelTask(taskId);
                throw new Exception("Task is timeout!");
            }else {
                throw new Exception("Invalid Task status!");
            }
        } else {
            throw new Exception("Invalid status code!");
        }
    }

}
